### PR TITLE
Allow downloads from a non-collection resolver

### PR DESCRIPTION
**Pull request by mickael9**
I noticed downloads don't work when results come from a regular (non-collection) resolver, so here is a small fix for that.

### CODE REVIEW

**Comment by dschmidt** [comment]
If we want this at all, it should go through `getStreamUrl` on the resolver too.. Where do you need this?

**Comment by mickael9** [comment]
@dschmidt Well, I wanted that feature for my resolver

And yes you can trigger it from the GUI, it shows the button if the resolver provides a `downloadUrls` but clicking on it doesn't do anything because `m_result->resolvedByCollection()` is null

Anyway, I made some changes, how about that?

**Comment by dschmidt** [comment]
Yes, better. :+1: 

Some thoughts tho:
1) `Collection` inherits `Resolver` (by now), we don't need the if-statement anymore -
just make sure `Resolver` implements the correct interface

2) I'm not a fan of changing the result parameters to QStrings, we lose context here
I see why you did this, the download format is not clear unlike the result's url. 
Let's leave `getStreamUrl` as it is and add `getDownloadUrl(const result_ptr&, DownloadFormat)` to the `Resolver` baseclass.

What do you think?

**Comment by mickael9** [comment]
1) Oh, okay. There's a default implementation in `Resolver.cpp` too

2)
* What's the `result_ptr` for? `DownloadFormat` has the URL in it
* Does that mean adding a `getDownloadUrl()` to the scripts themselves or does it still call `getStreamUrl` behind the scenes (that would be a bit confusing) ?

**Comment by dschmidt** [comment]
For completeness ;-) Who knows, maybe we need it later.. or we could check the result belongs to the collection/resolver at all. I just like not having to change APIs all over the place later on, if we could have done it "right" from the beginning on.

Yes, I would add it to the scripts as well - to the Resolver and Collection base classes in javascript - just underneath the getStreamUrl functions

**Comment by mickael9** [comment]
I was thinking about moving `DownloadFormat` into `Resolver` rather than `DownloadJob` (to avoid having to include `DownloadJob.h` from `Resolver.h`).

Should I move `DownloadFormat` into or outside the Tomahawk namespace?

What's the rationale for having things in the `Tomahawk` namespace vs  toplevel? (say `Tomahawk::ScriptJob` vs `::DownloadJob`)

**Comment by dschmidt** [comment]
Can you elaborate more on "moving DownloadFormat into Resolver rather than DownloadJob"?

The rationale is .. that we haven't been strict at any point. Actually anything in libtomahawk should be in the namespace, it's just easy to break especially runtime signal and slot connections when moving things around thus we kinda avoided moving everything in one rush.

**Comment by mickael9** [comment]
On the namespace thing, wouldn't using new-style `connect()` solve that problem ? I haven't seen those in the code.

Anyway, I restarted from scratch and did as you requested.
Let me know if you want me to make more changes :smiley:

**Comment by dschmidt** [comment]
Will look tomorrow, thanks already!

File: src/libtomahawk/DownloadJob.cpp [hunk 1]
```diff
@@ -206,22 +206,13 @@ DownloadJob::download()
...
     setState( Running );
 
-    if ( m_result->resolvedByCollection() )
-    {
-        Tomahawk::ScriptCollection* collection = qobject_cast<Tomahawk::ScriptCollection*>( m_result->resolvedByCollection().data() );
-        if ( collection )
-        {
-            QVariantMap arguments;
-            arguments[ "url" ] = m_format.url;
+    Tomahawk::ScriptJob *job = m_result->resolvedBy()->getDownloadUrl( m_result, m_format );
```
Comment (dschmidt) [inline src/libtomahawk/DownloadJob.cpp]: I think you need to check for resolvedBy() not being null

**Comment by mickael9** [comment]
There we go!

**Comment by dschmidt** [comment]
Totally forgot about this one ... great work, I'm sorry for taking so long :-(

**Pull request merged**